### ISSUE TITLE
fix: audit MD001/MD002/MD005/MD014/MD024/MD025/MD026/MD029/MD031/MD032/MD040 for missing frontmatter line-offset

### ISSUE THREAD

**Issue by joshrotenberg** [joshrotenberg/mdbook-lint#500]
## Summary

Eleven standard rules pair `document.node_position()` with direct `document.lines` indexing (or with line arithmetic) without adding `document.frontmatter_ast_offset(ast)`. This is the same bug class already fixed for MD041/MD022/MD007 (#406, PR #430) and MD004 (#441, PR #445).

comrak folds leading YAML frontmatter into a `FrontMatter` node and renumbers every following node as if the document began at line 1. A rule that takes an AST line and uses it against the real source is therefore off by the height of the frontmatter block.

This audit covers the `standard` ruleset's `node_position` users only. Every unfixed one is affected; the list below is complete for that surface.

## Reproduction

All fixtures below share this shape: frontmatter on lines 1-3, blank line 4, body starting on line 5.

```markdown
---
a: 1
---

# One

### Three
```

```
$ mdbook-lint lint --enable MD001 md001.md
error[MD001]: Expected heading level 2 (max 2) but got level 3: Three
  --> md001.md:3:1
     |
   3 | ---
     | ^^^ heading-increment
```

The heading is on line 7. The rule reports line 3, which is the closing frontmatter delimiter.

## Failure mode 1: autofix destroys the frontmatter

The reported position is also the position the `Fix` is anchored to, so `--fix` rewrites a frontmatter delimiter and leaves the real violation in place.

MD001, before and after `--fix`:

| Before | After |
| --- | --- |
| `1  ---` | `1  ---` |
| `2  a: 1` | `2  a: 1` |
| `3  ---` | `3  ## Three` |
| `4  ` | `4  ` |
| `5  # One` | `5  ` |
| `6  ` | `6  # One` |
| `7  ### Three` | `7  ` |
| | `8  ### Three` |

The closing `---` is gone, so the document no longer has frontmatter, and `### Three` is unchanged.

MD026 is worse because it lands on line 1:

```markdown
---
a: 1
---

# Hello!
```

After `mdbook-lint lint --enable MD026 --fix`:

```
1  Hello
2  
3  a: 1
4  ---
5  
6  # Hello!
```

The opening delimiter is replaced by the "fixed" heading text. MD025 behaves the same way, replacing the closing `---` with `## Two`.

## Failure mode 2: silent false negative

MD005 and MD029 read the offending line out of `document.lines` and derive the violation from its content. With frontmatter present they read the frontmatter lines instead, find nothing to complain about, and emit no violation at all.

MD005, identical body, with and without frontmatter:

```markdown
- a
 - b
- c
```

```
$ mdbook-lint lint --enable MD005 md005-plain.md
warning[MD005]: List item indentation inconsistent: expected 0 spaces, found 1
  --> md005-plain.md:2:1

$ mdbook-lint lint --enable MD005 md005.md      # same body, 4 lines of frontmatter prepended
No issues found
```

MD029 (`1.` / `3.` / `5.`) behaves identically: two warnings without frontmatter, zero with.

## Affected rules

Paths are relative to `crates/mdbook-lint-rulesets/src/standard/`.

| Rule | `node_position` | `document.lines` use | Observed | Expected | Notes |
| --- | --- | --- | --- | --- | --- |
| MD001 | md001.rs:101 | md001.rs:118 | line 3 | line 7 | `--fix` overwrites closing `---` |
| MD002 | md002.rs:84 | md002.rs:99 | not reproducible via CLI | | rule is `RuleMetadata::deprecated` (superseded by MD041); code path is identical to MD001 |
| MD005 | md005.rs:66 | md005.rs:67 | no violation | 1 warning | false negative |
| MD014 | md014.rs:63 | `base_line + line_idx + 1` | line 2 | line 6 | |
| MD024 | md024.rs:102, 184 | md024.rs:117, 201 | line 3 | line 7 | "first occurrence at line 1" should be line 5 |
| MD025 | md025.rs:78 | md025.rs:90 | line 3 | line 7 | "first at line 1" should be line 5; `--fix` overwrites closing `---` |
| MD026 | md026.rs:147 | md026.rs:171 | line 1 | line 5 | `--fix` overwrites opening `---` |
| MD029 | md029.rs:119 | md029.rs:120, 161 | no violation | 2 warnings | false negative |
| MD031 | md031.rs:48 | md031.rs:59, 67, 92, 96, 124, 134, 139, 151, 165 | line 2 (both) | line 6 | |
| MD032 | md032.rs:48 | md032.rs:98, 102, 152, 162, 167, 187 | line 2 | line 6 | also reads `sourcepos` directly at md032.rs:187 |
| MD040 | md040.rs:45 | none | line 3 | line 8 | report-only, so no fix corruption; fixture has 5 lines of frontmatter |

## Fix

Follow the pattern in md022.rs:47:

```rust
let frontmatter_offset = document.frontmatter_ast_offset(ast);
// ...
let line = line + frontmatter_offset;
```

The offset must be applied before the line is used for anything: the reported position, the `Fix` coordinates, and any `document.lines` read. MD031 and MD032 pass the line into helper functions (`has_blank_line_before`, `has_blank_line_after`, and the fence/list-extent scans), so the offset has to be applied at the call site or threaded through.

## Not yet audited

- `mdbook/` uses `node_position` in mdbook001.rs, mdbook002.rs, mdbook004.rs, mdbook006.rs. Not checked here.
- `adr/` and `content/` use no `node_position`, but seven ADR rules read `sourcepos` directly: adr001.rs, adr007.rs, adr009.rs, adr014.rs, adr015.rs, adr016.rs, adr017.rs. ADR documents commonly carry frontmatter, so this surface is worth a pass of its own.
- A further fourteen standard rules read `sourcepos` directly rather than going through `node_position`, and were outside this audit: MD003, MD007, MD009, MD023, MD034, MD042, MD043, MD045, MD046, MD048, MD051, MD056, MD058, MD059. (MD007 already handles frontmatter; the rest are unverified.)

## Suggested test convention

Each fix should add a case that lints the same body twice, once plain and once with frontmatter prepended, and asserts the violation lines differ by exactly the frontmatter height. That catches both failure modes, including the silent one.

**Comment by joshrotenberg** [comment]
ready: Slicing this audit. Slice 1 (MD005 + MD029, the silent false negatives) is in draft PR #501; remaining slices are MD001/MD002/MD024/MD025/MD026 and MD014/MD031/MD032/MD040.
